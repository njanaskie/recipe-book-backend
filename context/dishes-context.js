import React from 'react';

const DishesContext = React.createContext()

export { DishesContext as default }