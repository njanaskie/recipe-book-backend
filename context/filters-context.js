import React, { useContext, useReducer } from 'react';
import filtersReducer, { filtersReducerDefaultState } from '../src/reducers/filters'

const FiltersContext = React.createContext()

export const useFiltersContext = () => useContext(FiltersContext)

const FiltersProvider = ({ children }) => {
    const [filters, filtersDispatch] = useReducer(filtersReducer, filtersReducerDefaultState)

    return (
        <FiltersContext.Provider value={{ filters, filtersDispatch }}>
            {children}
        </FiltersContext.Provider>
    )
}

export { FiltersProvider, FiltersContext as default }