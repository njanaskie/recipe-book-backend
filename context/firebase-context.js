import React, { useEffect, useState, useContext, useReducer } from 'react';
import database, { firebase, googleAuthProvider } from '../src/firebase/firebase';
import { history } from '../src/routers/AppRouter';
import authReducer from '../src/reducers/auth'
import LoadingPage from '../src/components/LoadingPage'

const FirebaseContext = React.createContext()

export const useFirebaseContext = () => useContext(FirebaseContext)

const FirebaseProvider = ({ children }) => {
    const [auth, authDispatch] = useReducer(authReducer, [])
    // const authDispatch = useDispatch()
    const [user, setUser] = useState('')
    const [loading, setLoading] = useState(true)
    const [isAdmin, setIsAdmin] = useState()
    const [isGuest, setIsGuest] = useState()
    const email = process.env.GUEST_EMAIL
    const password = process.env.GUEST_PASSWORD

    const login = () => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(() => {
            authDispatch({ type: 'LOGIN' })
        })
    }

    const loginAsGuest = () => {
        firebase.auth().signInWithEmailAndPassword(email, password).then(() => {
            authDispatch({ type: 'LOGIN_AS_GUEST' })
        })
    }

    const logout = () => {
        firebase.auth().signOut().then(() => {
            authDispatch({ type: 'LOGOUT' })
        });
    }

    useEffect(() => {
        const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
            setUser(user)
            setLoading(false)
            setIsAdmin()
            setIsGuest()
        })

        return () => unsubscribe()
    }, [])

    useEffect(() => {
        if (user) {
            database.collection('users').doc(user.uid).get()
            .then((doc) => {
                const userData = doc.data()
                if (userData) {
                    if (userData.isAdmin === true) {
                        const isAdmin = userData.isAdmin
                        setIsAdmin(isAdmin)
                    } else if (userData.isGuest === true) {
                        const isGuest = userData.isGuest
                        setIsGuest(isGuest)
                    }
                }
            })
        }
    }, [user])

    if (loading) {
        return <LoadingPage />
    }
    // if (user) {
    //     if (history.location.pathname === '/') {
    //         history.push('/dashboard')
    //     }
    // } else {
    //     history.push('/');
    // }

    return (
        <FirebaseContext.Provider value={{ user, isAdmin, isGuest, login, loginAsGuest, logout }}>
            {children}
        </FirebaseContext.Provider>
    )
}

export { FirebaseProvider, FirebaseContext as default }