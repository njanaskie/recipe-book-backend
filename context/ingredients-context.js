import React, { useContext } from 'react';

const IngredientsContext = React.createContext()

export const useIngredientsContext = () => useContext(IngredientsContext)

export { IngredientsContext as default }