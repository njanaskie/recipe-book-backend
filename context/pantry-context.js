import React, { useContext } from 'react';

const PantryContext = React.createContext()

export const usePantryContext = () => useContext(PantryContext)

export { PantryContext as default }