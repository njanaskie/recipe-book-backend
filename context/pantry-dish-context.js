import React, { useContext } from 'react';

const PantryDishContext = React.createContext()

export const usePantryDishContext = () => useContext(PantryDishContext)

export { PantryDishContext as default }