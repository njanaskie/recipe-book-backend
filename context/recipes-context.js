import React, { useContext } from 'react';

const RecipesContext = React.createContext()

export const useRecipesContext = () => useContext(RecipesContext)

export { RecipesContext as default }