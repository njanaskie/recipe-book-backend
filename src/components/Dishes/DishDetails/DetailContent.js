import React, { useContext, useState, useEffect } from 'react'
import { Link } from 'react-router-dom';
import { ReactTinyLink } from 'react-tiny-link'
import { Tab, Button, Modal, Dropdown, Form } from 'semantic-ui-react'
import FirebaseContext from '../../../../context/firebase-context'
import RecipesContext from '../../../../context/recipes-context'
import useIngredients from '../../../hooks/useIngredients'
import useUserRecipes from '../../../hooks/useUserRecipes'
import database from '../../../firebase/firebase'
import UserRecipeItem from './UserRecipes/UserRecipeItem';
import UserRecipeForm from './UserRecipes/UserRecipeForm';
import AddUserRecipe from './UserRecipes/AddUserRecipe';
import EditUserRecipe from './UserRecipes/EditUserRecipe'
import UserRecipeModal from './UserRecipes/UserRecipeModal'

const DetailContent = ({ dish = {}, userRecipes = [], ingredients = [] }) => {
    const [isModalOpen, setIsModalOpen] = useState(false)
    const { isAdmin } = useContext(FirebaseContext)
    const { recipeDispatch } = useContext(RecipesContext)
    const { user } = useContext(FirebaseContext)

    const handleModalOpen = () => {
        setIsModalOpen(true)
    }

    const handleModalClose = () => {
        setIsModalOpen(false)
    }

    console.log(dish.recipes)
    console.log(userRecipes)

    const panes = [
        {
            menuItem: 'Suggested Recipes',
            render: () => <Tab.Pane>{dish.recipes && dish.recipes.map(recipe => recipe ? <ReactTinyLink key={recipe} url={recipe}>{recipe}</ReactTinyLink> : <p key={recipe}>No Suggested Recipes</p>)}</Tab.Pane>
        },
        {
            menuItem: 'My Saved Recipes',
            render: () => <Tab.Pane>{userRecipes.length > 0 ? userRecipes.map((recipe) => <UserRecipeItem key={recipe.id} recipe={recipe} dish={dish} isModalOpen={isModalOpen} handleModalOpen={handleModalOpen} handleModalClose={handleModalClose}/>) : <p>No Saved Recipes</p>}</Tab.Pane>
        },
    ]

    // const additionalIngredients = dish.keyIngredients && dish.keyIngredients.filter((keyIngredient) => {
    //     !ingredients.includes(keyIngredient)
    // })
    return (
        <div>
            <p>{dish.name}</p>
            <p>{dish.description}</p>
            <p>{dish.cuisine}</p>
            <p>{dish.type}</p>
            <h4>Key Ingredients</h4>
            {dish.keyIngredients && dish.keyIngredients.map(keyIngredient => <li key={keyIngredient}>{keyIngredient}</li>)}
            <h4>Optional Ingredients</h4>
            {dish.optionalIngredients && dish.optionalIngredients.map(optionalIngredient => <li key={optionalIngredient}>{optionalIngredient}</li>)}
            <div>
                <h3>Recipes</h3>
                <UserRecipeModal isModalOpen={isModalOpen} handleModalClose={handleModalClose}>
                    <AddUserRecipe dish={dish} handleModalClose={handleModalClose}/>
                </UserRecipeModal>
                <Button color='green' onClick={handleModalOpen}>Add Recipe</Button>
                <Tab panes={panes}/>
            </div>
            {isAdmin && <Link to={`/edit/dish/${dish.id}`}>Edit Dish</Link>}
        </div>
    )

}

export default DetailContent