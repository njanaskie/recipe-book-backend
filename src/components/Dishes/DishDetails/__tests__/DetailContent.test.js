import React from 'react'
import { shallow, mount } from 'enzyme'
import DetailContent from '../DetailContent'
import isAdmin from '../../../../tests/fixtures/admin'
import * as FirebaseContext from '../../../../../context/firebase-context'
import * as RecipesContext from '../../../../../context/recipes-context'


test('should render DetailContent correctly',() => {
    jest.spyOn(FirebaseContext, 'useFirebaseContext').mockImplementation(() => ({isAdmin: true}))
    const wrapper = shallow(<DetailContent />)
    expect(wrapper).toMatchSnapshot()
});