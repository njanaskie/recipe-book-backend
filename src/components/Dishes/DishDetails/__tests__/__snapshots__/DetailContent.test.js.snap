// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render DetailContent correctly 1`] = `
<div>
  <p />
  <p />
  <p />
  <p />
  <h4>
    Key Ingredients
  </h4>
  <h4>
    Optional Ingredients
  </h4>
  <div>
    <h3>
      Recipes
    </h3>
    <UserRecipeModal
      handleModalClose={[Function]}
      isModalOpen={false}
    >
      <AddUserRecipe
        dish={Object {}}
        handleModalClose={[Function]}
      />
    </UserRecipeModal>
    <Button
      as="button"
      color="green"
      onClick={[Function]}
    >
      Add Recipe
    </Button>
    <Tab
      grid={
        Object {
          "paneWidth": 12,
          "tabWidth": 4,
        }
      }
      menu={
        Object {
          "attached": true,
          "tabular": true,
        }
      }
      panes={
        Array [
          Object {
            "menuItem": "Suggested Recipes",
            "render": [Function],
          },
          Object {
            "menuItem": "My Saved Recipes",
            "render": [Function],
          },
        ]
      }
      renderActiveOnly={true}
    />
  </div>
  <[object Object]
    to="/edit/dish/undefined"
  >
    Edit Dish
  </[object Object]>
</div>
`;
